src/SAWScript/Proof.hs

```diff
@@ -421,41 +421,51 @@ simplifySequent sc ss (HypFocusedSequent (FB hs1 h hs2) gs) =
 
 
 hoistIfsInProp :: SharedContext -> Prop -> IO Prop
-hoistIfsInProp sc (Prop p) = do
-  let (args, body) = asPiList p
+hoistIfsInProp sc p = do
+  (ecs, body) <- unbindProp sc p
   body' <-
     case asEqTrue body of
       Just t -> pure t
       Nothing -> fail "hoistIfsInProp: expected EqTrue"
-  ecs <- traverse (\(nm, ty) -> scFreshEC sc nm ty) args
-  vars <- traverse (scExtCns sc) ecs
-  t0 <- instantiateVarList sc 0 (reverse vars) body'
-  t1 <- hoistIfs sc t0
+  t1 <- hoistIfs sc body'
   t2 <- scEqTrue sc t1
   t3 <- scGeneralizeExts sc ecs t2
   return (Prop t3)
 
+-- | Turn any leading Pi binders in the given prop into
+--   fresh ExtCns values, being careful to ensure that
+--   dependent types are properly substituted.
+unbindProp :: SharedContext -> Prop -> IO ([ExtCns Term], Term)
+unbindProp sc (Prop p0) = loop [] [] p0
```

Comment:
I wonder if this should be called something like `unbindAndFreshenProp`. After all, the key thing that distinguishes this from `asPiList` is how it freshens the pi binders that are split off, and the name `unbindProp` doesn't really convey this.